qiskit/utils/deprecate_pulse.py

```diff
@@ -0,0 +1,90 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Deprecation functions for Qiskit Pulse. To be removed in Qiskit 2.0.
+"""
+
+import warnings
+import functools
+
+from qiskit.utils.deprecation import deprecate_func, deprecate_arg
+
+
+def deprecate_pulse_func(func):
+    """Deprecation message for functions and classes"""
+    return deprecate_func(
+        since="1.3",
+        package_name="Qiskit",
+        removal_timeline="in Qiskit 2.0",
+        additional_msg="The entire Qiskit Pulse package is being deprecated "
+        "and will be moved to the Qiskit Dynamics repository: "
+        "https://github.com/qiskit-community/qiskit-dynamics",
+    )(func)
+
+
+def deprecate_pulse_dependency(*args, **kwargs):
+    """Deprecation message for functions and classes which use or depend on Pulse"""
+    decorator = deprecate_func(
+        since="1.3",
+        package_name="Qiskit",
+        removal_timeline="in Qiskit 2.0",
+        additional_msg="The entire Qiskit Pulse package is being deprecated "
+        "and this is a dependency on the package.",
+        **kwargs,
```

Comment:
I generally try to make all deprecation messages point to an alternative. From what I have understood, some dependencies might be moved to Qiskit Dynamics but some might not? Maybe we could say something like:

```suggestion
        additional_msg="The entire Qiskit Pulse package is being deprecated "
        "and this is a dependency on the package. The relevant components ",
         "for low-level pulse control have been moved to the Qiskit Dynamics ",
         "repository:  https://github.com/qiskit-community/qiskit-dynamics",
        **kwargs,
```

Reply:
Well, there will be no alternative in Qiskit for pulse and pulse-based things once Pulse is removed. So I think it makes sense to add your suggestion only for the things which will move with Pulse. And it can be refined to something like:  "The <X>component will be moved together with Qiskit Pulse to the Qiskit Dynamics repo...". For things which are not moving to Dynamics (e.g. the calibration passes) we can leave the original message or add something more explicit in the line of "Note that once removed in Qiskit 2.0, there will be no alternative functionality for this in Qiskit". What do you say?

Reply:
I think it makes sense

Reply:
Implemented in https://github.com/Qiskit/qiskit/pull/13164/commits/040119438c9f95b7a11a383a5d7aa8ce73fdad0c